extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java

```diff
@@ -97,14 +97,23 @@ void build(
         }
 
         for (AggregatedDataSourceBuildTimeConfigBuildItem aggregatedDataSourceBuildTimeConfig : aggregatedDataSourceBuildTimeConfigs) {
-            validateBuildTimeConfig(aggregatedDataSourceBuildTimeConfig);
 
+            String tracingDriver = null;
             if (aggregatedDataSourceBuildTimeConfig.getJdbcConfig().tracing) {
-                reflectiveClass
-                        .produce(new ReflectiveClassBuildItem(true, false,
-                                DataSources.TRACING_DRIVER_CLASSNAME));
+                if (capabilities.isPresent(Capability.OPENTELEMETRY_TRACER)) {
+                    tracingDriver = DataSources.OTEL_DRIVER_CLASSNAME;
+                } else if (capabilities.isPresent(Capability.SMALLRYE_OPENTRACING)) {
+                    tracingDriver = DataSources.TRACING_DRIVER_CLASSNAME;
+                } else {
+                    throw new ConfigurationException(
+                            "JDBC instrumentation has been enabled with configuration property 'quarkus.datasource.jdbc.tracing', but neither Quarkus OpenTelemetry nor SmallRye OpenTracing are present");
```

Comment:
I think we need tests on this logic ate the extension level and not only on IT tests.

Reply:
That logic will be gone when we go with datasource way, so I didn't add unit tests for it.